modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/io/PageIO.java

```diff
@@ -258,6 +258,9 @@
     /** */
     public static final short T_MARKER_PAGE = 33;
 
+    /** */
+    public static final short T_INDEX_META = 34;
```

Comment:
Can you please explain why you needed new IO type instead of new version for existing type? You only add two fields, right? I don't get it then.

Reply:
Yes, we adding two int fields into two pages.

We have 2 types of memory pages:
T_META (index partition)
T_PART_META (other partition)

We want to add 2 ints into each of them, but don't forget that IO for T_PART_META extends IO for T_META.

In my first approach I added 2 ints into T_META, and added PagePartitionMetaIOV3 for T_PART_META.
But since PagePartitionMetaIO "extends" PageMetaIO all offsets in PagePartitionMetaIO was shifted and there was a problem with inheritance (for example in PagePartitionMetaIO/V2 new methods were visible, all offsets was redefined in PagePartitionMetaIOV3), you can imagine what it looked like in commit with the second approach https://github.com/apache/ignite/pull/7941/commits/036b931bd40e2ffbd519779fde86fae7343da23e

With current approach we simply create the new IO that is only used for the "index" partition.

If you can suggest a better approach how for storing 2 extra int's itn the index meta page without losing binary compatibility, please describe.

p.s. I explained this change in wiki https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=95652384#TDE.Phase3.Cachekeyrotation.-PageMetaIOandPagePartitionMetaIO

Reply:
I suggest class PageIndexIO being just T_META with version 2, see no problems with this approach. Just as now, two different classes will add two independent sets of new integers, that's fine. The only thing that's changed is that PageIndexIO will have the same type as its base class.

Types should be immutable while pages are alive. This saves us from a lot of troubles, I wouldn't abandon this tradition. That's exactly what versions are for.

BTW your current code has bugs, you don't mark meta page as dirty in "getOrAllocateCacheMetas" while upgrading, please fix it.

Reply:
If I understood you correctly, you suggest to add PageMetaIOV2 (instead of PageIndexMetaIO)?
PagePartitionMeta will continue to extend PageMetaIO(not v2), is this correct?

Reply:
Yes, this is correct. Naming of the classes doesn't bother me too much so I was fine with "Indexing" instead of V2 (although that page is also referred as "cache group meta page" sometimes, like in PageMemoryEx#metaPageId, for example).

I think I understand the core problem, inheritance between MetaIO and PartitionMetaIO should have been replaced with extension of the same abstract superclass back in the day. But with current faulty design people just get confused :(

Reply:
I want to agree on naming for this change:

Type: T_META
Io: MetaPageIOV2 _(extends MetaPageIO)_

Wal record: MetaPageUpdate**Index**DataRecord 
Wal record type: **INDEX**_META_PAGE_DELTA_RECORD

is it ok?